### DISCUSSION TITLE
Direction-agnostic placement values

### DISCUSSION THREAD

**Discussion by TrevorRice** [floating-ui/floating-ui#1549] [Ideas]
Hi all! 👋

I'm curious — as support for languange-agnostic tools on the web increases (see [CSS logical properties](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Logical_Properties)), has any thought been given to making Floating UI's `Placement` values more "logical" rather than "physical"?

One thing I've noticed with most positioning libraries is the use of physical "left" and "right" values. Suppose I had a tooltip that needed to placed after a button. In a left-to-right language, I'd simply use "right". However, in a right-to-left language, it would need to be placed to the "left". If these values were direction-agnostic, consumers wouldn't need to write custom logic to flip their placement depending on the `dir` attribute.

For instance, Zendesk's [Garden](https://garden.zendesk.com/components/tooltip#placement) uses `top`, `topStart`, `topEnd`, `end`, `endTop`, `endBottom`, `bottom`, `bottomStart`, `bottomEnd`, `start`, `startTop`, and `startBottom` for their `Tooltip` positioning.

Of course, this is only [direction](https://developer.mozilla.org/en-US/docs/Web/CSS/direction)-agnostic — I think further consideration would need to be given to [writing modes](https://developer.mozilla.org/en-US/docs/Web/CSS/writing-mode).

Amazing work on Floating UI by the way!

**Comment by atomiks** [comment]
Yeah there's an umbrella issue to make it work more logically than physically: https://github.com/floating-ui/floating-ui/issues/1538. I actually just did that work tonight to do this and it'll be in the next release.

However we're going to use the following API: `${physicalSide}-${logicalAlignment}` by default.

Reasoning:

- `inline-start` and `block-start` etc are kind of confusing to most devs right now. It took me a while to "parse" how they work.
- Adding an alignment to the side makes the string confusing: `inline-start-end`. We'd need to split it into two options but this complicates a lot of things.

So only the `alignment` will be logical, leaving `side` to be physical.  If you want to make the `side` logical, rather than only the `alignment`, you'll need to use a custom middleware.

**Reply by TrevorRice** [reply]
Apologies — I skimmed through #1538 but somehow completely missed that section 😅

However we're going to use the following API: ${physicalSide}-${logicalAlignment} by default.

> Reasoning:
> 
> * inline-start and block-start etc are kind of confusing to most devs right now. It took me a while to "parse" how they work.
> * Adding an alignment to the side makes the string confusing: inline-start-end. We'd need to split it into two options but this complicates a lot of things.
> 
> So only the alignment will be logical, leaving side to be physical. If you want to make the side logical, rather than only the alignment, you'll need to use a custom middleware.

Cheers! I think this is a great improvement.

I appreciate the quick follow-up!

**Reply by atomiks** [reply]
I've been reading up on the new/upcoming/potential CSS anchoring properties and a space-separated list instead could work. The main issue is bundle/filesize for this library as this is also something an external function could technically handle.

Such syntax would handle left/right sides in RTL (rather than only alignments) — the issue is custom arrows needing to know the physical side, but we could make the returned data purely physical...

This is similar to CSS syntax:


```js
computePosition(reference, floating, {
  placement: 'block-end start',
}).then(({side, alignment}) => {
  // side => bottom in both RTL and LTR
  // alignment => left in LTR, right in RTL
});
```

```js
computePosition(reference, floating, {
  placement: 'inline-end end',
}).then(({side, alignment}) => {
  // side => right in LTR, left in RTL
  // alignment => bottom in both
});
```

Of course `writing-mode` is another different thing to handle.
